src-java/blue-green/src/main/java/org/openkilda/bluegreen/kafka/interceptors/VersioningProducerInterceptor.java

```diff
@@ -0,0 +1,96 @@
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.bluegreen.kafka.interceptors;
+
+import static org.openkilda.bluegreen.kafka.Utils.MESSAGE_VERSION_HEADER;
+import static org.openkilda.bluegreen.kafka.Utils.PRODUCER_COMPONENT_NAME_PROPERTY;
+import static org.openkilda.bluegreen.kafka.Utils.PRODUCER_RUN_ID_PROPERTY;
+import static org.openkilda.bluegreen.kafka.Utils.PRODUCER_ZOOKEEPER_CONNECTION_STRING_PROPERTY;
+import static org.openkilda.bluegreen.kafka.Utils.getValue;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.clients.producer.ProducerInterceptor;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+
+import java.time.Instant;
+import java.util.Map;
+
+@Slf4j
+public class VersioningProducerInterceptor<K, V> extends VersioningInterceptorBase
+        implements ProducerInterceptor<K, V> {
+
+    public VersioningProducerInterceptor() {
+        log.debug("Initializing VersioningProducerInterceptor");
+    }
+
+    @Override
+    public ProducerRecord<K, V> onSend(ProducerRecord<K, V> record) {
+        if (watchDog == null) {
+            // try to reconnect
+            initWatchDog();
+        }
+
+        if (record.headers().headers(MESSAGE_VERSION_HEADER).iterator().hasNext()) {
+            log.info("Kafka record {} already has header {}: {}",
+                    record, MESSAGE_VERSION_HEADER,
+                    new String(record.headers().headers(MESSAGE_VERSION_HEADER).iterator().next().value()));
+            record.headers().remove(MESSAGE_VERSION_HEADER);
+        }
+
+        if (version == null) {
+            if (isVersionIsNotTimeoutPassed()) {
+                // We will write this log every 60 seconds to do not spam in logs
+                log.warn("Messaging version is not set for component {} with id {}. Skip record {}",
+                        componentName, runId, record);
+                versionIsNotSetTimestamp = Instant.now();
+            }
+            // We can't return null record. We will return record without version.
+            // Such records wouldn't be read by Kafka Consumer with VersioningConsumerInterceptor
+            return record;
+        }
+
+        record.headers().add(MESSAGE_VERSION_HEADER, version);
+        return record;
+    }
+
+    @Override
+    public void onAcknowledgement(RecordMetadata metadata, Exception exception) {
+        // nothing to do
+    }
+
+    public void close() {
+        watchDog.unsubscribe(this);
+        watchDog.close();
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        connectionString = getValue(configs, PRODUCER_ZOOKEEPER_CONNECTION_STRING_PROPERTY, String.class);
```

Comment:
Is it safe to throw `IllegalArgumentException` from here?

Reply:
If we throw IllegalArgumentException for here running of a component stops. It needed for fast failing. Kilda wouldn't work without these config options